prospector/commit_preprocessor/preprocessor.py

```diff
@@ -83,6 +88,78 @@ def extract_cve_references(text: str) -> "list[str]":
     return [result.group(0) for result in re.finditer(r"CVE-\d{4}-\d{4,8}:?", text)]
 
 
+def extract_commit_ids_from_website_recursively(
+    url: str, max_depth: int, repository_url: str
+) -> Set[str]:
+    """
+    Extract referenced commit ids from a website and from other
+    referenced websites recursively
+
+    @param: url url of a website
+    @param: max_depth how deep we should go in the hyperlink reference tree
+    @param: repository_url url of the github page of project, like https://github.com/flatpak/flatpak
+    """
+    urls_to_visit = {url}
+    visited_urls = set()
+    commit_ids = set()
+    for depth in range(max_depth):
+        new_urls_to_visit = set()
+        for url in urls_to_visit:
+            if url in visited_urls:
+                continue
+            visited_urls.add(url)
+            found_urls = extract_referenced_urls_from_url(url)
+            if found_urls:
+                relevant_urls = found_urls - urls_to_visit - visited_urls
+                for referenced_url in set(relevant_urls):
+                    commit_id = extract_referenced_commit_id_from_url(
+                        referenced_url, repository_url
+                    )
+                    if commit_id:
+                        commit_ids.add(commit_id)
+                        relevant_urls.remove(referenced_url)
+                new_urls_to_visit |= relevant_urls
+        urls_to_visit = new_urls_to_visit
+    return commit_ids
+
+
+def extract_referenced_urls_from_html(html: str) -> Set[str]:
+    """
+    Extract referenced URLs from html code
+    """
+    soup = BeautifulSoup(html, "html.parser")
+    return {link.get("href") for link in soup.find_all("a") if link.get("href") != None}
+
+
+def extract_referenced_urls_from_url(url: str) -> Optional[Set[str]]:
+    """
+    Extract referenced URLs from website
+    """
+    parsed = urlparse(url)
+    if not bool(parsed.netloc) or not bool(parsed.scheme):
+        return None
+    response = requests.get(url)
+    if response:
+        return extract_referenced_urls_from_html(response.text)
+    return None
+
+
+def extract_referenced_commit_id_from_url(
```

Comment:
Not sure where/how we can use this; why not return repo+commit_id? (basically a Commit obj)

Reply:
To make this useful in practice, it should take an arbitrary url and determine if it points to a git repo + commit, in which case it should return these two elements.

Returning just the commit id has limited use IMHO.